controllers/cluster_controller.go

```diff
@@ -278,70 +301,231 @@ func (r *ClusterReconciler) upgradeControlPlane(ctx context.Context, cluster *ca
+	if expectedMachineImage != currentMachineImage || expectedK8sVersion != *machineDeployment.Spec.Template.Spec.Version {
+		nodesWillBeRolled = true
+
+		logger.Info("MachineDeployment needs to be upgraded and nodes will be rolled")
+		logger.Info("Cloning current referenced infrastructure template changing its machine image", "currentMachineTemplate", machineDeployment.Spec.Template.Spec.InfrastructureRef.Name)
+
+		// Clone infrastructure machine template to new object.
+		newInfrastructureMachineTemplate, err := r.CloneTemplate(ctx, infraMachineTemplateReference, machineDeployment.Namespace, machineDeployment.Name)
+		if err != nil {
+			return false, errors.Wrapf(err, "failed to build infrastructure template from %q", infraMachineTemplateReference.Name)
+		}
+
+		// Update machine template machine image to upgrade k8s or the OS.
+		if err := unstructured.SetNestedField(newInfrastructureMachineTemplate.Object, expectedMachineImage, strings.Split(KindToMachineImagePath[infraMachineTemplateReference.Kind], ".")...); err != nil {
+			return false, errors.Wrapf(err, "failed to set infrastructure template %q image to %q", infraMachineTemplateReference.Name, expectedMachineImage)
+		}
+
+		// Create cloned object.
+		if err := r.Client.Create(ctx, newInfrastructureMachineTemplate); err != nil {
+			return false, errors.Wrapf(err, "failed to clone infrastructure template from %q to %q used by KubeadmControlPlane %q", infraMachineTemplateReference.Name, newInfrastructureMachineTemplate.GetName(), machineDeployment.Name)
+		}
+
+		// When upgrading the k8s cluster we need to update
+		// * Name of the secret used as source for the k8s provider config file
+		// * Reference to the infrastructure machine template.
+		// * K8s version in the control plane object.
+		kubeadmconfigtemplate := &cabpk.KubeadmConfigTemplate{}
+		err = r.Client.Get(ctx, client.ObjectKey{Namespace: machineDeployment.Namespace, Name: machineDeployment.Spec.Template.Spec.Bootstrap.ConfigRef.Name}, kubeadmconfigtemplate)
+		if err != nil {
+			return false, errors.Wrapf(err, "failed to retrieve machine deployment bootstrap KubeadmConfigTemplate %q", machineDeployment.Spec.Template.Spec.Bootstrap.ConfigRef.Name)
+		}
+
+		for _, file := range kubeadmconfigtemplate.Spec.Template.Spec.Files {
+			if file.ContentFrom.Secret.Name == fmt.Sprintf(KindToK8sProviderFile[newInfrastructureMachineTemplate.GetKind()], machineDeployment.Spec.Template.Spec.InfrastructureRef.Name) {
+				file.ContentFrom.Secret.Name = fmt.Sprintf(KindToK8sProviderFile[newInfrastructureMachineTemplate.GetKind()], newInfrastructureMachineTemplate.GetName())
+			}
+		}
+		err = r.Client.Update(ctx, kubeadmconfigtemplate)
```

Comment:
Is this the same `KubeadmConfigTemplate` used by the "old" infra machine template, or is it a copy too? If it's the same, do you think there's a risk it could break existing machines?

Reply:
No, nothing specific, as I'm not even sure if existing machines can be affected in any way by changing the templates. I actually wanted to check if you can think of one, maybe with a rollout being triggered externally (like with `clusterctl alpha rollout restart`)?

Reply:
Let's keep it this way and I'll try to break it with further tests